src/ring_theory/dedekind_domain/finite_adele_ring.lean

```diff
@@ -0,0 +1,264 @@
+  of `R`.
+- `finite_adele_ring` : The finite adèle ring of `R`, defined as the restricted product
+  `Π'_v K_v`.
+
+## Implementation notes
+We are only interested on Dedekind domains of Krull dimension 1 (i.e., not fields). If `R` is a
+field, its finite adèle ring is just defined to be the trivial ring.
+
+## References
+* [J.W.S. Cassels, A. Frölich, *Algebraic Number Theory*][cassels1967algebraic]
+
+## Tags
+finite adèle ring, dedekind domain
+-/
+
+noncomputable theory
+open function set is_dedekind_domain
+
+variables (R : Type) (K : Type) [comm_ring R] [is_domain R] [is_dedekind_domain R] [field K]
+  [algebra R K] [is_fraction_ring R K] (v : height_one_spectrum R)
+
+/-- The product of all `adic_completion_integers`, where `v` runs over the maximal ideals of `R`. -/
+def prod_completions_integers := (Π (v : height_one_spectrum R), v.adic_completion_integers K)
+local notation `R_hat` := prod_completions_integers
+
+instance : comm_ring (R_hat R K) := pi.comm_ring
+
+instance : topological_space (R_hat R K) := Pi.topological_space
+
+instance prod_completions_integers.inhabited : inhabited (prod_completions_integers R K) := ⟨0⟩
+
+/-- The product of all `adic_completion`, where `v` runs over the maximal ideals of `R`. -/
+def prod_completions := (Π (v : height_one_spectrum R), v.adic_completion K)
+
+local notation `K_hat` := prod_completions
+
+instance : comm_ring (K_hat R K) := pi.comm_ring
+instance : ring (K_hat R K) := infer_instance
+instance : topological_space (K_hat R K) := Pi.topological_space
+instance : topological_ring (K_hat R K) :=
+(infer_instance : topological_ring (Π (v : height_one_spectrum R), v.adic_completion K))
+
+instance prod_completions.inhabited : inhabited (prod_completions R K) := ⟨0⟩
+
+/-- The diagonal inclusion `r ↦ (r)_v` of `R` in `R_hat`. -/
+def diag_inj : R → (R_hat R K) := λ r v, inj_adic_completion_integers R K v r
+
+lemma diag_inj.map_one : diag_inj R K 1 = 1 :=
+by { rw diag_inj, simp_rw ring_hom.map_one, refl }
+
+lemma diag_inj.map_mul (x y : R): diag_inj R K (x*y) = (diag_inj R K x) * (diag_inj R K y) :=
+by { rw diag_inj, ext v, apply congr_arg _ (ring_hom.map_mul _ _ _) }
+
+namespace prod_completions_integers
+
+noncomputable! instance : has_coe (R_hat R K) (K_hat R K) := { coe := λ x v, x v }
+
+lemma coe_apply (x : R_hat R K) (v : height_one_spectrum R) : (x : K_hat R K) v = ↑(x v) := rfl
+
+/-- The inclusion of `R_hat` in `K_hat` is a homomorphism of additive monoids. -/
+noncomputable! def coe.add_monoid_hom : add_monoid_hom (R_hat R K) (K_hat R K) :=
+{ to_fun    := coe,
+  map_zero' := rfl,
+  map_add'  := λ x y, by { ext v, simp only [coe_apply, pi.add_apply, subring.coe_add] }}
+
+lemma coe.add_monoid_hom_apply (x : R_hat R K) (v : height_one_spectrum R) :
+  (coe.add_monoid_hom R K) x v = x v := rfl
+
+/-- The inclusion of `R_hat` in `K_hat` is a ring homomorphism. -/
+noncomputable! def coe.ring_hom : ring_hom (R_hat R K) (K_hat R K)  :=
```

Comment:
I don't know whether `K_hat` has an `algebra R` instance (but it should I think): in this case this should be an `alg_hom`.

Reply:
I have used `coe.ring_hom` to define the algebra instance `algebra (R_hat R K) (K_hat R K)`, and I have added the corresponding `coe.alg_hom`.